### PR TITLE
Show a warning prior to opening external links

### LINKED ISSUE #3496: External Links warning
> I've been thinking about this idea for a while, and I finally made a concept to show to you.
## Concept
Make the option (and the feature) to choose if people want a warning for external links
## "I don't know why this can be useful, explain why this should be added"
Many times can happen that you open a link by mistake, or you don't think that it will open the browser (confusing it for one of the in-game "pages"), and this can be "stressful" sometimes (especially for slow devices). Also, this feature is present in other application (like Discord), or even in other games, so it's very common.
## Concept
- It's just a Photoshop with the _confirm elimination_ warning design edited (and with a different icon, 2 designs), but it can explain better the idea:

external link warning concept style 1

external link warning concept style 2

- And this is the _disable_ option on the **Online** section, because I presume that some people don't want the warning:

external link warning option concept

### PR DESCRIPTION
**Pull request by UselessToucan**
Closes #3496 

external-link-warning

About using safe browsing which was mentioned [here](https://github.com/ppy/osu/issues/3496#issuecomment-424377671): Google has [free](https://developers.google.com/safe-browsing/v4/pricing) safebrowsing service, but its basic [usage limits](https://developers.google.com/safe-browsing/v4/usage-limits) are suitable only for development and testing. In order to use it production, peppy will have to request additional quota.

### CODE REVIEW

**Comment by R0slaN** [comment]
may i request you to add a "trust this domain in the future" button?
i'd imagine this feature would get annoying quickly without one.

File: osu.Game/OsuGame.cs [hunk 1]
```diff
@@ -103,6 +104,9 @@ private Intro intro
 
         private readonly List<OverlayContainer> overlays = new List<OverlayContainer>();
 
+        private ExternalLinkOpener externalLinkOpener;
+        public void OpenUrlExternally(string url) => externalLinkOpener.OpenUrlExternally(url);
```
Comment (smoogipoo) [inline osu.Game/OsuGame.cs]: Please put methods below the constructor and `load()` methods. Those that are currently not there are also misplaced but you don't have to change them for now.

File: osu.Game/Overlays/Chat/ExternalLinkDialog.cs [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+namespace osu.Game.Overlays.Chat
+{
+    public class ExternalLinkDialog : PopupDialog
+    {
+        public ExternalLinkDialog(string url, Action openExternalLinkAction)
+        {
+            BodyText = url;
+
+            Icon = FontAwesome.fa_warning;
+            HeaderText = "Confirm opening external link";
```
Comment (smoogipoo) [inline osu.Game/Overlays/Chat/ExternalLinkDialog.cs]: Maybe "Are you sure you want to open the following?"

File: osu.Game/Online/Chat/ExternalLinkOpener.cs [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+        private void load(GameHost host, DialogOverlay dialogOverlay, OsuConfigManager config)
+        {
+            this.host = host;
+            this.dialogOverlay = dialogOverlay;
+            warnAboutOpeningExternal = config.GetBindable<bool>(OsuSetting.WarnAboutOpeningExternalLink);
+        }
+
+        public void OpenUrlExternally(string url)
+        {
+            void externalAction() => host.OpenUrlExternally(url);
```
Comment (smoogipoo) [inline osu.Game/Online/Chat/ExternalLinkOpener.cs]: Really don't need a nested function here, just in-line it in the two places.

File: osu.Game/Configuration/OsuConfigManager.cs [hunk 4]
```diff
@@ -148,6 +150,7 @@ public enum OsuSetting
         BeatmapSkins,
         BeatmapHitsounds,
         IncreaseFirstObjectVisibility,
-        ScoreDisplayMode
+        ScoreDisplayMode,
+        WarnAboutOpeningExternalLink
```
Comment (smoogipoo) [inline osu.Game/Configuration/OsuConfigManager.cs]: `ExternalLinkWarning`

File: osu.Game/Online/Chat/ExternalLinkOpener.cs [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+using osu.Game.Overlays;
+using osu.Game.Overlays.Chat;
+
+namespace osu.Game.Online.Chat
+{
+    public class ExternalLinkOpener : Component
+    {
+        private GameHost host;
+        private DialogOverlay dialogOverlay;
+        private Bindable<bool> warnAboutOpeningExternal;
```
Comment (smoogipoo) [inline osu.Game/Online/Chat/ExternalLinkOpener.cs]: `externalLinkWarning`

**Review by smoogipoo — CHANGES_REQUESTED** [review]
In general seems fine, just needs some cleanup

**Review by smoogipoo — APPROVED** [review]

**Pull request merged**